Makefile: fix generating coverage when specifiying TESTPKGS

christarazi opened the pull request:
See commit msg for details. This should help surface coverage information when developing on individual packages, and such.

joestringer reviewed (changes requested):
Assuming this works as intended and doesn't adversely impact the Coveralls integration in CI, I think the only other piece is that presumably the `COVERPKG` variable is no longer needed and can be removed from the git tree.

Alternatively when running locally you *can* specify `COVERPKG=...` to choose the package to apply coverage for, although while this works it's not a great workflow so if this PR solves the problem more generally I'd rather get rid of it.

christarazi committed:
Makefile: don't run duplicate cmd in TESTPKGS_EVAL

Previously, `TESTPKGS_EVAL` would shell out to `$(GO_LIST)` when that
had already been executed. This commit uses the output of the previous
command instead of running the command twice.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Makefile: fix generating coverage with TESTPKGS

This commit fixes the problem where the package that is passed to
`-coverpkg` is not full package path, meaning having the prefix of
"github.com/cilium/cilium". For example, if I wanted to test the `eni`
package, I would need to pass "github.com/cilium/cilium/pkg/aws/eni".
Before this commit, the value passed is "pkg/aws/eni".

Before:
$ make SKIP_KVSTORES=true TESTPKGS=pkg/rand unit-tests
...
for pkg in github.com/cilium/cilium/pkg/rand; do \
        go test -mod=vendor ... $pkg -test.v -timeout 360s -coverprofile=coverage.out -covermode=count -coverpkg pkg/rand || exit 1; \
        tail -n +2 coverage.out >> coverage-all-tmp.out; \
done
warning: no packages being tested depend on matches for pattern pkg/rand
=== RUN   Test
OK: 1 passed
--- PASS: Test (0.00s)
=== RUN   TestIntn
--- PASS: TestIntn (0.00s)
PASS
coverage: [no statements]
ok      github.com/cilium/cilium/pkg/rand       0.003s  coverage: [no statements]
make generate-cov
make[1]: Entering directory '/home/chris/code/cilium/cilium'
dirname: missing operand
Try 'dirname --help' for more information.
\# Remove generated code from coverage
grep -Ev '(^github.com/cilium/cilium/api/v1)|(generated.deepcopy.go)|(^github.com/cilium/cilium/pkg/k8s/client/)' \
        coverage-all-tmp.out > coverage-all.out
go tool cover -html=coverage-all.out -o=coverage-all.html
rm coverage.out coverage-all-tmp.out
make[1]: Leaving directory '/home/chris/code/cilium/cilium'
make stop-kvstores
...

After:
$ make SKIP_KVSTORES=true TESTPKGS=pkg/rand unit-tests
...
for pkg in github.com/cilium/cilium/pkg/rand; do \
        go test -mod=vendor ... $pkg -test.v -timeout 360s \
        -coverpkg $pkg -coverprofile=coverage.out -covermode=count || exit 1; \
        tail -n +2 coverage.out >> coverage-all-tmp.out; \
done
=== RUN   Test
OK: 1 passed
--- PASS: Test (0.00s)
=== RUN   TestIntn
--- PASS: TestIntn (0.00s)
PASS
coverage: 36.1% of statements in github.com/cilium/cilium/pkg/rand
ok      github.com/cilium/cilium/pkg/rand       0.006s  coverage: 36.1% of statements in github.com/cilium/cilium/pkg/rand
make generate-cov
make[1]: Entering directory '/home/chris/code/cilium/cilium'
dirname: missing operand
Try 'dirname --help' for more information.
\# Remove generated code from coverage
grep -Ev '(^github.com/cilium/cilium/api/v1)|(generated.deepcopy.go)|(^github.com/cilium/cilium/pkg/k8s/client/)' \
        coverage-all-tmp.out > coverage-all.out
go tool cover -html=coverage-all.out -o=coverage-all.html
rm coverage.out coverage-all-tmp.out
make[1]: Leaving directory '/home/chris/code/cilium/cilium'
make stop-kvstores
...

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on Makefile:
```diff
@@ -161,7 +161,8 @@ endif
 	for pkg in $(patsubst %,github.com/cilium/cilium/%,$(TESTPKGS)); do \
-		$(GO_TEST) $(TEST_UNITTEST_LDFLAGS) $$pkg $(GOTEST_BASE) $(GOTEST_COVER_OPTS) || exit 1; \
+		$(GO_TEST) $(TEST_UNITTEST_LDFLAGS) $$pkg $(GOTEST_BASE) $(GOTEST_COVER_OPTS) || \
+		exit 1; \
```
nit, is this just a stylistic change? Elsewhere in the Makefile we use a slightly different style to try to highlight the conditional:

```suggestion
		$(GO_TEST) $(TEST_UNITTEST_LDFLAGS) $$pkg $(GOTEST_BASE) $(GOTEST_COVER_OPTS) \
		|| exit 1; \
```

christarazi replied:
Addressed

joestringer reviewed (changes requested):
Minor style nit, otherwise looks good. Good to see a filesystem iteration getting removed here too, that should speed up builds.
